src/lib.rs

```diff
@@ -1038,12 +1071,26 @@ impl<'a> Cmd<'a> {
                 stdin.flush()
             }));
         }
-        let out_res = child.wait_with_output();
-        let err_res = io_thread.map(|it| it.join().unwrap());
-        let output = out_res.map_err(|err| Error::new_cmd_io(self, err))?;
-        if let Some(err_res) = err_res {
-            err_res.map_err(|err| Error::new_cmd_stdin(self, err))?;
+
+        let output = if let Some(duration) = timeout {
+            let (sender, receiver) = mpsc::channel();
+            thread::spawn(move || {
+                let output = child.wait_with_output();
+                let _ = sender.send(output);
+            });
```

Comment:
This is not quiet right --- we are "leaking" a thread handle here. As we are not manually `.join`ing the thread, this means this thread can continue running past the point where `output_impl` returns, which violates structured concurrency. 

We should enforce the invariant that, by the time `output_impl` returns, the child thread is `.joined`. 

We should also proactively `.kill` the child process when the timeout happens. My understanding is that the current behavior is that the function retuns with an error after the timeout elapses, but the actual child process continues to execute, which is not what we want here.

Reply:
Yes, you're absolutely right. Thank you for pointing it out. I'll find an appropriate place to `.join` and `.kill` this thread to prevent the "leaking".

Reply:
However, it seems that I hit the wall:
```rust
        let out_res = if let Some(timeout) = self.data.timeout {
            let (tx, rx) = mpsc::channel();
            let handle = thread::spawn(move || {
                let output = child.wait_with_output();
                let _ = tx.send(output);
            });
            handle.join().unwrap();
            match rx.recv_timeout(timeout) {
                Ok(output) => output,
                Err(err) => {
                    // FIXME: Kill the child, borrow of moved value: `child`
                    // child.kill();
                    return Err(Error::new_timeout(self, err));
                }
            }
        } else {
            child.wait_with_output()
        };
```
Since `.wait_with_output` need the child's ownership, when spawned, I can't use it to call `.kill` anymore. I tried Arc<Mutex> way, but it still failed:
```
error[E0507]: cannot move out of dereference of MutexGuard<'_, Child>
```
@matklad Do you have any suggestion for this situation?

Reply:
Huh, that's nasty! This sadly might require us to essentially re-implement `wait_with_output`, at least for the case where there's a timeout

Reply:
@Binlogo Saw this issue and thought I would mention this is basically the entire purpose of [process\_control](https://docs.rs/process_control/latest/process_control/).

Reply:
thanks for your information, it seems fit this situation, i'll take a look.